From b446ecfeb3b4fc77ab31212b33a0635d749cae53 Mon Sep 17 00:00:00 2001
From: "Liu,Dalei" <daleix.liu@intel.com>
Date: Fri, 12 Jun 2015 14:19:37 +0800
Subject: [PATCH 423/462] DW9718: Add power on flag

Add a flag for power on to avoid i2c error when VCM turned off.

Change-Id: I6d12b37fa6dc56f41efe886b3a7d40c01839d141
Tracked-On:https://jira01.devtools.intel.com/browse/IMINAN-39149
Signed-off-by: Liu,Dalei <daleix.liu@intel.com>
Reviewed-on: https://icggerrit.ir.intel.com/60142
Approver: Yunliang Ding <yunliang.ding@intel.com>
Reviewed-by: Yunliang Ding <yunliang.ding@intel.com>
Validation-Android: Wei Tao <wei.tao@intel.com>
Integrator: WeiX Meng <weix.meng@intel.com>
Build: IMIN (SC)
Build: WeiX Meng <weix.meng@intel.com>
Reviewed-by: WeiX Meng <weix.meng@intel.com>
Maintainer: WeiX Meng <weix.meng@intel.com>
---
 .../camera/drivers/media/i2c/imx/dw9718.c              | 18 +++++++++++++++++-
 .../camera/drivers/media/i2c/imx/dw9718.h              |  1 +
 2 files changed, 18 insertions(+), 1 deletion(-)

diff --git a/drivers/external_drivers/camera/drivers/media/i2c/imx/dw9718.c b/drivers/external_drivers/camera/drivers/media/i2c/imx/dw9718.c
index 225b477..9e04787 100644
--- a/drivers/external_drivers/camera/drivers/media/i2c/imx/dw9718.c
+++ b/drivers/external_drivers/camera/drivers/media/i2c/imx/dw9718.c
@@ -104,6 +104,9 @@ int dw9718_vcm_power_up(struct v4l2_subdev *sd)
 	int i;
 	int step = DW9718_CLICK_REDUCTION_STEP;
 
+	if (dw9718_dev.power_on)
+		return 0;
+
 	/* Enable power */
 	ret = dw9718_dev.platform_data->power_ctrl(sd, 1);
 	if (ret)
@@ -164,6 +167,7 @@ int dw9718_vcm_power_up(struct v4l2_subdev *sd)
 		return ret;
 
 	dw9718_dev.initialized = true;
+	dw9718_dev.power_on = 1;
 
 	return 0;
 
@@ -180,6 +184,9 @@ int dw9718_vcm_power_down(struct v4l2_subdev *sd)
 	int i;
 	int step = DW9718_CLICK_REDUCTION_STEP;
 
+	if (!dw9718_dev.power_on)
+		return 0;
+
 	/* Minimize the click sounds from the lens during power down */
 	i = min(dw9718_dev.focus, DW9718_DEFAULT_FOCUS_POSITION) - step;
 	while (i >= DW9718_LENS_MOVE_POSITION) {
@@ -197,7 +204,15 @@ int dw9718_vcm_power_down(struct v4l2_subdev *sd)
 		dev_err(&client->dev, "%s: write DW9718_PD to 1 failed\n",
 				__func__);
 
-	return dw9718_dev.platform_data->power_ctrl(sd, 0);
+	ret =  dw9718_dev.platform_data->power_ctrl(sd, 0);
+	if (ret) {
+		dev_err(&client->dev, "%s power_ctrl failed\n",
+				__func__);
+		return ret;
+	}
+	dw9718_dev.power_on = 0;
+
+	return 0;
 }
 
 int dw9718_q_focus_status(struct v4l2_subdev *sd, s32 *value)
@@ -251,5 +266,6 @@ int dw9718_vcm_init(struct v4l2_subdev *sd)
 {
 	dw9718_dev.platform_data = camera_get_af_platform_data();
 	dw9718_dev.focus = DW9718_DEFAULT_FOCUS_POSITION;
+	dw9718_dev.power_on = 0;
 	return (NULL == dw9718_dev.platform_data) ? -ENODEV : 0;
 }
diff --git a/drivers/external_drivers/camera/drivers/media/i2c/imx/dw9718.h b/drivers/external_drivers/camera/drivers/media/i2c/imx/dw9718.h
index b878df5..2726c7b 100644
--- a/drivers/external_drivers/camera/drivers/media/i2c/imx/dw9718.h
+++ b/drivers/external_drivers/camera/drivers/media/i2c/imx/dw9718.h
@@ -36,6 +36,7 @@ struct dw9718_device {
 	struct timespec focus_time;	/* Time when focus was last time set */
 	__u8 buffer[4];			/* Used for i2c transactions */
 	const struct camera_af_platform_data *platform_data;
+	__u8 power_on;
 };
 
 #define DW9718_MAX_FOCUS_POS	1023
-- 
1.9.1

